### PR TITLE
Add support to multiple package sources in YAML files

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by mopsfelder — COMMENTED** [review]
I see `dest`, `dest_dir`, and `directory` variable names to represent the same thing.

I also noted that some functions have a default value for directory variable, e.g. `directory=''`.  Shall we specify a default value for all the other directory variables?

How about adopting a unified name for such kind of variable, be it `dest`, `dest_dir`, or `directory`?

File: lib/mockbuilder.py [hunk 1]
```diff
@@ -98,10 +101,17 @@ def _prepare(self, package):
...
+        if package.sources:
+            archive_to_build_dir = partial(package_source.archive,
+                                           directory=self.build_dir)
+            archived_sources = map(archive_to_build_dir, package.sources)
+            package.sources = archived_sources
+            self.archive = self.build_dir
+        elif package.repository:
+                self.archive = package.repository.archive(package.expects_source,
+                                                                  package.commit_id,
+                                                                  self.build_dir)
```
Comment (ghost) [inline lib/mockbuilder.py]: Minor: weird indentation.

File: lib/package.py [hunk 2]
```diff
@@ -123,29 +131,23 @@ def _load(self):
...
-        # a custom file and regex.
         version = self.package_data.get('version', {})
         self.version_file_regex = (version.get('file'),
                                    version.get('regex'))
 
-        # NOTE(maurosr): Unfortunately some of the packages we build
-        # depend on a gziped file which changes according to the build
-        # version so we need to get that name somehow, grep the
-        # specfile would be uglier imho.
+        # (ltadeu): These are all being kept for compatibility reasons. They are
```
Comment (ghost) [inline lib/package.py]: Unrelated to this patch: I don't see the point of leaving the author's name in the comments. If we really need to find out who wrote them, git blame can be used.

File: lib/package_source.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    repo.checkout(commit_id or branch)
+    source['git']['repo'] = repo
+    return source
+
+
+def _url_download(source, dest_dir):
+    """
+    Downloads a file from URL [source] and returns a source dict.
+    """
+    file_name = source['url']['src'].rsplit('/', 1)[1]
```
Comment (ghost) [inline lib/package_source.py]: I think using os.path.basename is a little bit more legible.

File: lib/package_source.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    cmd = 'hg archive -t tgz "{}"'.format(archive_file)
+    utils.run_command(cmd, cwd=source['hg']['dest'])
+
+    source['hg']['archive'] = archive_file
+    return source
+
+
+def _url_archive(source, directory):
+    dest = source['url']['dest']
+    file_name = dest.rsplit('/', 1)[1]
```
Comment (ghost) [inline lib/package_source.py]: I think using os.path.basename is a little bit more legible.

File: lib/package_source.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    archive_path = os.path.join(directory, file_name)
+
+    
+    cmd = 'mv {} {}'.format(dest, archive_path)
```
Comment (ghost) [inline lib/package_source.py]: Use either shutil.move or os.rename

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@mopsfelder @olavphibm fixed those issues.

All functions now use `directory` variable.

File: lib/mockbuilder.py [hunk 6]
```diff
@@ -12,15 +12,19 @@
...
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from functools import partial
+
 
 import datetime
 import logging
 import os
 import shutil
+import sys
```
Comment (ghost) [inline lib/mockbuilder.py]: Unused import `sys`

**Review by ghost — CHANGES_REQUESTED** [review]

File: lib/mockbuilder.py (hunk 6 above)
Comment (ghost) [reply lib/mockbuilder.py]: Damn pylint didn't complain about this on my laptop :rage:

Comment (ghost) [reply lib/mockbuilder.py]: Are you using the latest pylint version and the .pylintrc located in this repo?

File: lib/package_source.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    if commit_id is None and branch is None:
+        raise ValueError('invalid hg source dict: missing both `commit_id` '
+                         'and `branch`')
+
+    command += 'clone "{}" -r "{}" "{}"'.format(hg_source['src'],
+                                                hg_source['branch'],
+                                                dest)
+    proc = Popen(command.split(' '), stderr=PIPE)
+    proc.communicate()
```
Comment (gyribeiro) [inline lib/package_source.py]: It is preferable use lib.utils.run_command to execute external commands instead of use directly Popen.

**Comment by olavphibm** [comment]
start pylint

File: lib/package_source.py (hunk 7 above)
Comment (ghost) [reply lib/package_source.py]: Hmm, I didn't remember we had that. I'll change the code to use our `run_command`.

**Review by ghost — APPROVED** [review]

**Review by gyribeiro — APPROVED** [review]

**Comment by olavphibm** [comment]
start tests

**Comment by ghost** [comment]
This closes #111

**Pull request merged**